### PR TITLE
Akka.Streams update 2.4.14

### LINKED ISSUE #2600: Backport changes from Streams 2.4.14
## 2.4.14 (29.10.2016 - 18.11.2016)

- [x] - [FlattenMerge should propagate outer attributes](https://github.com/akka/akka/pull/21805)
- [ ] - [Move compression tests over from akka-http + fix for ByteStringParser](https://github.com/akka/akka/pull/21872)
- [ ] - [Move (de)compression helpers to akka-stream module ](https://github.com/akka/akka/pull/21409)
- [ ] - [Failing tcp spec fix](https://github.com/akka/akka/pull/21685)
- [x] - [expose `within` for both TestSubscriber and TestPublisher](https://github.com/akka/akka/pull/21851)
- [ ] - [Fix ByteStringParser backpressure](https://github.com/akka/akka/pull/21853)
- [x] - [Correct errors from InputStreamSinkStage when first read times out](https://github.com/akka/akka/pull/21821)
- [x] - [Fix race in BroadcastHub shutdown and new registrations](https://github.com/akka/akka/pull/21855)
- [x] - [use SimpleLinearGraphStage to reduce the boilerplate code](https://github.com/akka/akka/pull/21829)
- [x] - [rewrite Buffer as a GraphStage](https://github.com/akka/akka/pull/21635)
- [x] - [Fixed error handling behaviour of Fold and Reduce](https://github.com/akka/akka/pull/21797)
- [ ] - [Close Tls connection if Inbound closed before receiving peer's close_notify](https://github.com/akka/akka/pull/21786)
- [ ] - [foldAsync test using mapAsync](https://github.com/akka/akka/pull/21720)
- [ ] - [move compression tests](https://github.com/akka/akka/pull/21876)

### PR DESCRIPTION
**Pull request by Silv3rcircl3**
closes #2600

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@Silv3rcircl3 missing an API approval apparently, since bringing in the latest changes

File: src/core/Akka.Streams/ActorMaterializer.cs [hunk 1]
```diff
@@ -145,6 +145,15 @@ private static ActorSystem ActorSystemOf(IActorRefFactory context)
         /// <summary>
         /// TBD
         /// </summary>
+        /// <typeparam name="TMat">TBD</typeparam>
+        /// <param name="runnable">TBD</param>
+        /// <param name="initialAttributes">TBD</param>
+        /// <returns>TBD</returns>
+        public abstract TMat Materialize<TMat>(IGraph<ClosedShape, TMat> runnable, Attributes initialAttributes);
```
Comment (alexvaluyskiy) [inline src/core/Akka.Streams/ActorMaterializer.cs]: Comment is missed 
```
 This method interprets the given `Flow` description and creates the running stream using an explicitly provided `Attributes` as top level attributes. The result can be highly implementation specific, ranging from local actor chains to remote-deployed processing networks.
```

**Comment by alexvaluyskiy** [comment]
Reviewed `use SimpleLinearGraphStage to reduce the boilerplate code`
`DelayInitial`, `Detacher`, `SkipWhile`, `Recover`, `Intersperse`, `LimitWeighted` did not converted to `SimpleLinearGraphStage`

Why we have changed naming `DropWhile -> SkipWhile`, but we still have `Drop` stage?

**Comment by alexvaluyskiy** [comment]
Reviewed `rewrite Buffer as a GraphStage`. Did not find this change
```
-  def buffer(size: Int, overflowStrategy: OverflowStrategy): Repr[Out] = andThen(Buffer(size, overflowStrategy))
 +  def buffer(size: Int, overflowStrategy: OverflowStrategy): Repr[Out] = via(fusing.Buffer(size, overflowStrategy))
```

**Comment by alexvaluyskiy** [comment]
What with these features?
[Failing tcp spec fix](https://github.com/akka/akka/pull/21685)
[Close Tls connection if Inbound closed before receiving peer's close_notify](https://github.com/akka/akka/pull/21786)

**Comment by Silv3rcircl3** [comment]
@alexvaluyskiy 
- Failing tcp spec fix is in included in the Akka IO update
- Close Tls connection, we haven't ported the TLS part since at the time even the akka remove had no TLS support, so I'll create an issue for this since we now have support for it. 
- Fix ByteStringParser, this one is only used for GzipDecompressor and was initially implemented in akka http so I think we can skip it until we port akka http

**Comment by Silv3rcircl3** [comment]
the buffer changes are [here](https://github.com/akkadotnet/akka.net/pull/2694/commits/d265abb653bf4578dce433eddaa30c2640ef74f6#diff-f903bfc50e7f4d462c55c4d5dd7e9749R1126)

**Comment by alexvaluyskiy** [comment]
Is TSL support should use remoting, not Akka IO?

**Comment by Silv3rcircl3** [comment]
After a short look at the TLS api I don't think they use either akka.remote nor akka.io , it's a bidi stage where TLS encoded bytes come in and decoded bytes go out without any direkt communication to the network layer.

**Pull request merged**